File: Contest/OmegaUp/5/E.py
from sys import stdin, stdout
import re

check = {}
inside = {}

class Cont:
    pad = 0
    spac = 0
    w = 0
    h = 0
    area = -1
    isContainer = False
    isVertical = False
    
    def __init__(self, name):
        self.name = name

    def setContainer(self, flag):
        self.isContainer = True
        self.isVertical = flag

    def setWindow(self, W, H):
        self.w = W
        self.h = H
        self.area = W*H

    def setPadding(self, P):
        self.pad = P

    def setSpacing(self, Sp):
        self.spac = Sp

def calcArea(name):
    if name in inside:
        for c in inside[name]:
            if check.get(c).area == -1:
                calcArea(c)

            if check.get(name).isVertical:
                check.get(name).h += check.get(c).h
                check.get(name).w = max(check.get(name).w, check.get(c).w)
            else:
                check.get(name).w += check.get(c).w
                check.get(name).h = max(check.get(name).h, check.get(c).h)

    if check.get(name).isContainer and inside.get(name) != None:
        check.get(name).w += 2 * check.get(name).pad
        check.get(name).h += 2 * check.get(name).pad
        if check.get(name).isVertical:
            check.get(name).h += check.get(name).spac * (len(inside[name])-1)
        else:
            check.get(name).w += check.get(name).spac * (len(inside[name])-1)

    check.get(name).area = check.get(name).w * check.get(name).h

n = int(stdin.readline())

for _ in range(n):
    s = stdin.readline().strip('\n')
    arr = re.split("[.(), ]",s)
    if s.find('.') != -1:
        aux = check.get(arr[0])
        if arr[1] == "add":
            if arr[0] in inside:
                inside[arr[0]].append(arr[2])
            else:
                inside[arr[0]] = [arr[2]]
        elif arr[1] == "setPadding":
            aux.setPadding(int(arr[2]))
        else:
            aux.setSpacing(int(arr[2]))
        check[arr[0]] = aux
    else:
        aux = Cont(arr[1])
        if len(arr) == 2:
            aux.setContainer(arr[0] == "VerticalContainer")
        else:
            aux.setWindow(int(arr[2]), int(arr[3]))
        check[arr[1]] = aux

for c in check.values():
    if c.area == -1:
        calcArea(c.name)

for c in sorted(check):
    stdout.write("{} {} {}\n".format(c, check.get(c).w, check.get(c).h))

File: Contest/OmegaUp/5/E.cpp
#include <bits/stdc++.h>

using namespace std;

typedef __int128_t Long;

ostream&
operator<<(ostream& dest, __int128_t value)
{
	ostream::sentry s(dest);
	if(s){
		__uint128_t tmp = value < 0 ? -value : value;
		char buffer[128];
		char *d = end(buffer);
		do{
			--d;
			*d = "012345679"[tmp%10];
			tmp /= 10;
		}while(tmp != 0);
		if(value < 0){
			--d;
			*d = '-';
		}
		int len = end(buffer)-d;
		if(dest.rdbuf()->sputn(d,len) != len){
			dest.setstate(ios_base::badbit);
		}
	}
	return dest;
}

__int128_t max(__int128_t &a, __int128_t &b){
	return (a > b ? a : b);
}

struct cont{
	string name;
	Long pad, spac, w, h, area;
	bool isContainer, isVertical;
	vector<cont*> a;

	cont(string S, bool F): name(S), pad(0), spac(0), isContainer(true), isVertical(F), area(-1), w(0), h(0){}
	cont(string S, Long W, Long H): name(S), pad(0), spac(0), isContainer(false), area(W*H), w(W), h(H){}
};

void calcArea(cont *c){
	Long w = c->w, h = c->h;
#ifdef DEBUG
	cout << c->name << " " << w << ":" << h << " -> ";
#endif
	for(auto &con : c->a){
		if(con->area == -1ll)
			calcArea(con);
#ifdef DEBUG
		cout << "add(" << con->w << "," << con->h << ")" << con->isVertical << " ";
#endif
		if(c->isVertical){
			h += con->h;
			w = max(w, con->w);
		}
		else{
			w += con->w;
			h = max(h, con->h);
		}
	}
	
#ifdef DEBUG
	cout << w << ":" << h << " -> ";
#endif
	if(c->isContainer && !(c->a.empty())){
		w += (c->pad) << ((Long)1ll);
		h += (c->pad) << ((Long)1ll);
		if(c->isVertical)
			h += (c->spac)*((c->a.size())-(Long)1ll);
		else
			w += (c->spac)*((c->a.size())-(Long)1ll);
	}
#ifdef DEBUG
	cout << w << ":" << h << endl;
#endif
	c->h = h;
	c->w = w;
	c->area = w*h;
}

int main(){
	int n;
	cin >> n;cin.ignore();

	string s;
	map<string, cont*> check;
	for(int i{0}; i < n; i++){
		getline(cin,s);
		if(s.find(".") != string::npos){
			stringstream ss(s);
			string n1, f, n2;
			getline(ss, n1, '.');
			getline(ss, f, '(');
			getline(ss, n2, ')');
			if(f == "add")
				check[n1]->a.push_back(check[n2]);
			else if(f == "setPadding")
				check[n1]->pad = stoll(n2);
			else
				check[n1]->spac = stoll(n2);
		}else{
			stringstream ss(s);
			string type, aux, name, w, h;
			getline(ss, type, ' ');
			getline(ss, aux, ' ');
			if(aux.find("(") != string::npos){
				stringstream ss1(aux);
				getline(ss1, name, '(');
				getline(ss1, w, ',');
				getline(ss1, h, ')');
				check[name] = new cont(name, stoll(w), stoll(h));
			}else{
				name = aux;
				if(type.find("Vertical") != string::npos)
					check[name] = new cont(name, true);
				else
					check[name] = new cont(name, false);
			}
		}
	}


	for(auto &p : check){
		if(p.second->area == -1){
			calcArea(p.second);
		}
	}

	for(auto &p : check){
		cout << p.first << " " << p.second->w << " " << p.second->h << endl;
	}
}
